### DISCUSSION TITLE
Reverse-Mode Autodifferentiation for MuJoCo Warp: Design and Roadmap

### DISCUSSION THREAD

**Discussion by mar-yan24** [google-deepmind/mujoco_warp#1242] [Ideas]
# Reverse-Mode Autodifferentiation for MuJoCo Warp: Design and Roadmap

## TLDR

- **Phase 1 (implemented, PR #1226):** Reverse-mode AD through smooth dynamics: kinematics, velocity, actuation, integration. This should allow most people to compute gradients of scalar losses like in `qpos`, `qvel`, `ctrl`, and other state bc of `wp.Tape`.
- **Phase 2 (workin on it right now):** Constraint solver gradients via implicit differentiation. Probably gonna implement it so it bypasses non-differentiable solver internals (e.g. convergence loops, tile Cholesky) by analytically differentiating the KKT conditions at convergence. AD-only protected in this implementation, where clones have a `requires_grad` check so non-AD paths pay zero.
- **Phase 3 (planned):** Collision pipeline gradients via smooth signed-distance proxies. Replaces discrete geometric queries (GJK/EPA, integer branching) with differentiable distance functions for each primitive type. Want to avoid discontinuities and have full grad flow.
- **Together, these three phases bring MJX-equivalent differentiable physics to the Warp/CUDA backend**. This is mainly gonna be a prototype that I implement, I hope the community will try it out a bit and let me know if there are any major performance issues or bugs.

### Selective `enable_backward` strategy

Warp generates adjoint, backward, kernels for every `@wp.kernel` in modules where `enable_backward=True`. Enabling this globally would increase compilation time across 100+ kernels (the concern raised in [discussion #993](https://github.com/google-deepmind/mujoco_warp/discussions/993)). Instead, we enable backward selectively on only the modules that participate in differentiable paths:

| Module | `enable_backward` | Gradient mechanism |
|---|---|---|
| `smooth.py` | `True` | Tape AD |
| `forward.py` | `True` | Tape AD |
| `passive.py` | `True` | Tape AD |
| `derivative.py` | `True` | Tape AD |
| `constraint.py` | `False` | (Phase 3 future) |
| `solver.py` | `False` | Custom adjoint (implicit diff) |
| `collision_*.py` | `False` | Custom adjoint (smooth proxies) |
| All others (sensor, render, ray, ...) | `False` | Not differentiated |

This generates adjoint kernels for about 30 smooth-dynamics kernels out of 100+ total. Warp caches compiled kernels, so the cost is one-time per kernel signature. Within the enabled modules, individual tile kernels still carry `enable_backward=False` overrides because cuSolverDx LTO compilation does not support adjoint generation.

---

## Phase 1: Smooth Dynamics (Implemented)

Phase 1 covers the contact-free forward dynamics path: `kinematics -> fwd_velocity -> fwd_actuation -> euler/rk4`.

### What's covered

- Kinematics: body positions, orientations, Jacobians, composite rigid body inertia
- Velocity propagation: bias forces, Coriolis, centripetal terms
- Actuation: actuator forces, gravity compensation, clamping
- Integration: Euler and RK4 time stepping
- Custom adjoint for `quat_integrate` to avoid gradient singularity at zero angular velocity

### Public API

```python
import mujoco_warp as mjw
import warp as wp

# Option 1: Convenience wrapper
d = mjw.make_diff_data(mjm)                    # make_data + enable_grad
tape = mjw.diff_step(m, d, my_loss_fn)         # tape + step + backward
grad_ctrl = d.ctrl.grad                        # read gradients

# Option 2: Direct tape usage
d = mjw.make_data(mjm)
mjw.enable_grad(d)                             # mark fields as requires_grad
tape = wp.Tape()
with tape:
    mjw.step(m, d)
    wp.launch(loss_kernel, dim=1, inputs=[d.xpos, target, loss])
tape.backward(loss=loss)
grad_ctrl = d.ctrl.grad
```

Key functions: `enable_grad`, `disable_grad`, `make_diff_data`, `diff_step`, `diff_forward`. Field lists: `SMOOTH_GRAD_FIELDS` (~40 Data fields), `SOLVER_GRAD_FIELDS` (constraint-related fields).

### Warp workarounds (summary)

There were some workarounds I had to implement when testing the AD capabilities native to Warp. I actually thought they were bugs at first but they're known limitations within the differentiability structure of Warp for performance reasons. Feel free to take a look at the recent closed issues in Warp but this note is just for anyone if you see workarounds for ad reverse loop and stuff lol.

---

### Why tape-based AD is blocked for the solver

The constraint solver, `solver.py`, is a Newton-type optimizer that finds constraint forces satisfying KKT conditions. Four properties make it incompatible with tape-based AD:

1. **Convergence-based iteration**: `wp.capture_while` runs until all worlds converge independently, producing variable-length traces that the tape cannot replay backward.
2. **Constraint activation discontinuity**: `pos < 0` branching in `constraint.py` can flip active constraints with small perturbations, changing the Jacobian structure discontinuously.
3. **Linesearch with conditional swaps**: Per-world, per-Newton-step branching that the backward pass cannot reconstruct.
4. **Tile Cholesky without adjoint support**: `wp.tile_cholesky` / `wp.tile_cholesky_solve` use cuSolverDx, which does not support LTO adjoint generation.

### The implicit differentiation approach

I am in the middle of doing some workarounds for this constraint solver but right now I am planning on implementing/lowkey already implementing implicit diff method. Instead of differentiating through solver iterations, maybe should instead differentiate the solver's stationarity condition at convergence.

### Implementation details

Already implemented a retained solver state, where the solver's Hessian, `solver_h`, holesky factor `solver_hfactor`, and reference force `solver_Jaref`, are stored on `Data` for reuse during the backward pass.

Gonna use the  `tape.record_func` to inject a callables into the tape's launch list. During `tape.backward() it goes into the right position in the reverse direction.

In theory, when the callable goes off, `d.qacc.grad` already contains `dL/dqacc` from the integration backward. The callable solves the adjoint linear system and seeds `d.qacc_smooth.grad` so the smooth dynamics backward can continue. But uh ill update here.

### Limitations with this

- **Newton solver only**: The CG solver path falls back to an identity adjoint
- **Fixed active set assumption**: Implicit differentiation assumes small perturbations don't change which constraints are active. Gradients are unreliable near activation boundaries. Unfortunately this is basically an inherent limitation shared by all implicit-diff approaches as far as I'm aware
- **No gradients through constraint construction**: this is cause the following phase is still planned

---

### Why the discrete collision code can't be differentiated

The collision pipeline, `collision_driver.py` + narrowphase modules, uses discrete geometric algorithms that are fundamentally non-differentiable:

- **Integer configuration encoding**: `cltype`, `clface`, `clcorner` are integers controlling branching with no gradient
- **Data-dependent contact count**: `wp.atomic_add(ncollision_out, ...)` makes output size depend on input values
- **GJK/EPA variable iteration**: Convergence depends on specific geometry pairs
- **SDF octree traversal**: Data-dependent paths with inner gradient descent loops
- **Broadphase sorting**: `wp.tile_sort` is a discrete permutation

Any advice on how to tackle these would be great before I get started.

---

## Roadmap

| Phase | Status | Description |
|---|---|---|
| **1: Smooth dynamics** | Done | Tape AD through kinematics, velocity, actuation, integration |
| **1.5: Conditional clone** | Done | Guard AD-only `wp.clone` calls with `requires_grad` checks |
| **2a: Solver state retention** | Done | Store Hessian/Cholesky factor on `Data` for backward pass |
| **2b-d: Implicit diff adjoint** | Done | `tape.record_func` injection, adjoint linear system solve, tests |
| **3a: Smooth distance (sphere/capsule/plane)** | Done | Analytically smooth primitives with natural Warp AD |
| **3b: Collision proxy adjoints** | Done | Custom adjoints mapping contact gradients to body transforms |
| **3c: Fixed-size contact buffers** | Testing | Differentiable contact buffer structure |
| **3d: Smooth distance (box/mesh/convex)** | Planned | More complex SDF formulations |
| **3e: End-to-end contact AD** | Planned | Full gradient chain validation with contacts |

**Comment by mar-yan24** [comment]
Was able to get cart-pole short horizon actor critic trained on my end using full mjwarp diff pipeline.
